openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java

```diff
@@ -0,0 +1,520 @@
+    return super.getVersionInternal(securityContext, id, version);
+  }
+
+  @POST
+  @Operation(
+      operationId = "createQuery",
+      summary = "Create a query",
+      tags = "query",
+      description = "Create a query under an existing entity.",
+      responses = {
+        @ApiResponse(
+            responseCode = "200",
+            description = "The query",
+            content =
+                @Content(
+                    mediaType = "application/json",
+                    schema = @Schema(implementation = QueryResource.QueryList.class))),
+        @ApiResponse(responseCode = "400", description = "Bad request")
+      })
+  public Response create(@Context UriInfo uriInfo, @Context SecurityContext securityContext, @Valid CreateQuery create)
+      throws IOException {
+    Query query = getQuery(create, securityContext.getUserPrincipal().getName());
+    return create(uriInfo, securityContext, query);
+  }
+
+  @PUT
+  @Operation(
+      operationId = "createOrUpdateQuery",
+      summary = "Create or update a query",
+      tags = "tables",
+      description = "Create a query, if it does not exist. If a query already exists, update the query.",
+      responses = {
+        @ApiResponse(
+            responseCode = "200",
+            description = "The query",
+            content = @Content(mediaType = "application/json", schema = @Schema(implementation = Query.class))),
+        @ApiResponse(responseCode = "400", description = "Bad request")
+      })
+  public Response createOrUpdate(
+      @Context UriInfo uriInfo, @Context SecurityContext securityContext, @Valid CreateQuery create)
+      throws IOException {
+    Query query = getQuery(create, securityContext.getUserPrincipal().getName());
+    return createOrUpdate(uriInfo, securityContext, query);
+  }
+
+  @PATCH
+  @Path("/{id}")
+  @Operation(
+      operationId = "patchQuery",
+      summary = "Update a query",
+      tags = "query",
+      description = "Update an existing query using JsonPatch.",
+      externalDocs = @ExternalDocumentation(description = "JsonPatch RFC", url = "https://tools.ietf.org/html/rfc6902"))
+  @Consumes(MediaType.APPLICATION_JSON_PATCH_JSON)
+  public Response patch(
+      @Context UriInfo uriInfo,
+      @Context SecurityContext securityContext,
+      @Parameter(description = "Id of the query", schema = @Schema(type = "UUID")) @PathParam("id") UUID id,
+      @RequestBody(
+              description = "JsonPatch with array of operations",
+              content =
+                  @Content(
+                      mediaType = MediaType.APPLICATION_JSON_PATCH_JSON,
+                      examples = {
+                        @ExampleObject("[" + "{op:remove, path:/a}," + "{op:add, path: /b, value: val}" + "]")
+                      }))
+          JsonPatch patch)
+      throws IOException {
+    return patchInternal(uriInfo, securityContext, id, patch);
+  }
+
+  @PUT
+  @Path("/{id}/followers")
+  @Operation(
```

Comment:
@harshach do we want followers for query?

Reply:
Queries can be added through Ingestion, and maybe through the UI as well in the near future, this will allow users who want to follow a frequently used query or a query that have a use for them , let me know if doesn't make sense we can remove it

Reply:
We can make following a separate step. Let's remove this for the time being (once you add an API we can't remove it, especially post 1.0).